src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -918,37 +922,69 @@ private boolean completed(Collection<SnapshotMetaData.ShardSnapshotStatus> shard
      * @param request update shard status request
      */
     private void innerUpdateSnapshotState(final UpdateIndexShardSnapshotStatusRequest request) {
+        logger.debug("received updated snapshot restore state [{}]", request);
+        updatedSnapshotStateQueue.add(request);
+
         clusterService.submitStateUpdateTask("update snapshot state", new ClusterStateUpdateTask() {
+            private final List<UpdateIndexShardSnapshotStatusRequest> drainedRequests = new ArrayList<>();
+
             @Override
             public ClusterState execute(ClusterState currentState) {
+
+                if (request.processed)
+                {
+                    return currentState;
+                }
+
+                updatedSnapshotStateQueue.drainTo(drainedRequests);
+
+                final int batchSize = drainedRequests.size();
+
+                // nothing to process (a previous event has processed it already)
+                if (batchSize == 0)
+                {
+                    return currentState;
+                }
+
                 MetaData metaData = currentState.metaData();
                 MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
                 SnapshotMetaData snapshots = metaData.custom(SnapshotMetaData.TYPE);
                 if (snapshots != null) {
-                    boolean changed = false;
-                    ArrayList<SnapshotMetaData.Entry> entries = newArrayList();
-                    for (SnapshotMetaData.Entry entry : snapshots.entries()) {
-                        if (entry.snapshotId().equals(request.snapshotId())) {
-                            HashMap<ShardId, ShardSnapshotStatus> shards = newHashMap(entry.shards());
-                            logger.trace("[{}] Updating shard [{}] with status [{}]", request.snapshotId(), request.shardId(), request.status().state());
-                            shards.put(request.shardId(), request.status());
-                            if (!completed(shards.values())) {
-                                entries.add(new SnapshotMetaData.Entry(entry, ImmutableMap.copyOf(shards)));
+                    int changedCount = 0;
+                    List<SnapshotMetaData.Entry> entries = snapshots.entries();
+                    
+                    for (int i = 0; i < batchSize; i++) {
+                        UpdateIndexShardSnapshotStatusRequest updateSnapshotState = drainedRequests.get(i);
+                        updateSnapshotState.processed = true;
+
+                        List<SnapshotMetaData.Entry> newEntries = newArrayList();
+                        for (SnapshotMetaData.Entry entry : entries) {
```

Comment:
It’s much better than what we have today, but I think we can optimize it a bit more if we will move this loop to the outer level. This way we can apply all shards with updated status and only then at the end check if the snapshot is completed or not and process it once. This way we don’t have to check all shards and rebuild the snapshot metadata over and over again. What do you think?